tensorflow/core/kernels/segment_reduction_ops_impl.h

```diff
@@ -957,6 +957,125 @@ class SparseSegmentSqrtNGradOp
                                                      true /*is_sqrtn*/) {}
 };
 
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+template <class Device, class T, class Tindex, bool has_num_segments>
+class SparseSegmentSumGpuOp : public AsyncOpKernel {
+ public:
+  explicit SparseSegmentSumGpuOp(OpKernelConstruction* context)
+      : AsyncOpKernel(context){};
+
+  void ComputeAsync(OpKernelContext* context, DoneCallback done) override {
+    const Tensor& input_data = context->input(0);
+    const Tensor& indices = context->input(1);
+    const Tensor& segment_ids = context->input(2);
+
+    OP_REQUIRES_ASYNC(context, TensorShapeUtils::IsVector(indices.shape()),
+                      errors::InvalidArgument("indices should be a vector."),
+                      done);
+    const int64 num_indices = indices.NumElements();
+
+    OP_REQUIRES_ASYNC(
+        context, TensorShapeUtils::IsVector(segment_ids.shape()),
+        errors::InvalidArgument("segment_ids should be a vector."), done);
+
+    OP_REQUIRES_ASYNC(
+        context, num_indices == segment_ids.NumElements(),
+        errors::InvalidArgument("indices and segment_ids should have"
+                                "same length."),
+        done);
+
+    ScratchSpace<Index> output_rows_host(context, 1, /* on_host */ true);
+    auto stream = context->op_device_context()->stream();
+
+    if (has_num_segments) {
+      const Tensor& num_segments = context->input(3);
+      se::DeviceMemoryBase num_segments_device(
+          const_cast<Tensor&>(num_segments).template flat<Index>().data()
+      );
+      OP_REQUIRES_ASYNC(
+          context,
+          stream
+              ->ThenMemcpy(output_rows_host.mutable_data(), num_segments_device, sizeof(Index)).ok(),
+          errors::Internal(
+              "SparseSegmentSumGpuOp: failed to copy num_segments to host."),
+          done);
+    } else {
+      se::DeviceMemoryBase last_segment_id_on_device(
+          const_cast<Tensor&>(segment_ids).template flat<Index>().data() +
+          (num_indices - 1));
```

Comment:
Are you sure the `const_cast` is needed?  I think you should be able to create and pass a `const DeviceMemoryBase` to `ThenMemcpy`.

Reply:
The pointer here is used to get the value of the last id in `segment_ids`. [here](https://github.com/tensorflow/tensorflow/pull/45085/commits/13c00d7463afd62c6a5425ad35aaea1e4c377d63#diff-d29812caca45335066ddacf8fb06afc416f1f557bc446068ebaa163495e0b03bR995)